rust/kernel/src/lib.rs

```diff
@@ -56,3 +58,24 @@ fn panic(_info: &PanicInfo) -> ! {
 
 #[global_allocator]
 static ALLOCATOR: allocator::KernelAllocator = allocator::KernelAllocator;
+
+/// Attempts to allocate memory for the given value using the global allocator. On success, the
+/// value is moved into it and returned to the caller wrapped in a Box.
+pub fn try_alloc<T>(value: T) -> KernelResult<Box<T>> {
+    let layout = core::alloc::Layout::for_value(&value);
+    let rptr = unsafe { alloc::alloc::alloc(layout) };
+    if rptr.is_null() {
+        return Err(Error::ENOMEM);
+    }
+    let ptr = rptr as *mut T;
+    unsafe {
+        ptr.write(value);
+        Ok(Box::from_raw(ptr))
+    }
```

Comment:
Based on how `Box` works, I think we want something a little smarter. Looking at:

- https://github.com/rust-lang/rust/blob/master/library/alloc/src/boxed.rs#L262
- https://github.com/rust-lang/rust/blob/master/library/alloc/src/boxed.rs#L292

We want something like:

```rs
        let layout = Layout::new::<mem::MaybeUninit<T>>();
        let ptr = alloc::alloc::alloc(layout);
        // Handle nul ptr here
        let mut b = unsafe { Box::from_raw(ptr.as_ptr()) };
        b.as_mut_ptr().write(value);
        b.assume_init()
```

Reply:
For my own benefit: do you see any UB in the code above?

I've updated the code as you suggested. In process of adding SAFETY comments, I noticed that we cannot call `alloc` with a zero-sized layout, so I added some code to take care of that.

I also don't call assume_init() because it requires some nightly feature. I just do the into_raw/from_raw dance.

PTAL.